### ISSUE TITLE
File::Find not descending into Joliet-filesystem

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3048]
Migrated from [rt.perl.org#5036](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5036) (status was 'resolved')

Searchable as RT5036$

**Comment by p5pRT** [comment]
### From torsten@sotlx2.sot.de

#### Created by scheck@bigfoot.de


When descending into a Joliet filesystem Find stops after the
first level\. Rockridge or ISO CDs are ok\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by prospector at Mon Aug  7 10:58:30 EDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.5-22smp, archname=i386-linux
    uname='linux porky.devel.redhat.com 2.2.5-22smp #1 smp wed jun 2 09:11:51 edt 1999 i686 unknown '
    config_args='-des -Doptimize=-O2 -march=i386 -mcpu=i686 -Dcc=gcc -Dcccdlflags=-fPIC -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dd_dosuid -Dd_semctl_semun -Di_db -Di_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Uuselargefiles'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=undef 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O2 -march=i386 -mcpu=i686', gccversion=2.96 20000731 (experimental)
    cppflags='-fno-strict-aliasing'
    ccflags ='-fno-strict-aliasing'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lcrypt
    libc=/lib/libc-2.1.92.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i386-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i386-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/torsten
    LANG=de_DE
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/torsten/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


Site configuration information for perl 5.00503:

Configured by root at Wed Feb  2 15:34:50 EST 2000.

Summary of my perl5 (5.0 patchlevel 5 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.2.5-22smp, archname=i386-linux
    uname='linux porky.devel.redhat.com 2.2.5-22smp #1 smp wed jun 2 09:11:51 edt 1999 i686 unknown '
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cc', optimize='-O2 -m486 -fno-strength-reduce', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    ccflags ='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    stdchar='char', d_stdstdio=undef, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lposix -lcrypt
    libc=, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.00503:
    /usr/lib/perl5/5.00503/i386-linux
    /usr/lib/perl5/5.00503
    /usr/lib/perl5/site_perl/5.005/i386-linux
    /usr/lib/perl5/site_perl/5.005
    .


Environment for perl 5.00503:
    HOME=/home/torsten
    LANG=de
    LANGUAGE (unset)
    LC_ALL=de_DE
    LD_LIBRARY_PATH=:/opt/SYBSasa6/lib
    LOGDIR (unset)
    PATH=.:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/opt/informix/bin:/opt/SYBSasa6/bin:/opt/kde/bin:/usr/local/bin:/mnt/d/data/projekt/mp3/bin:/usr/local/bin:/home/torsten/bin/vnc_x86_linux_2.0
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @floatingatoll

I don't have time to handle this until the weekend; figured I'd forward it
along and see if someone else helps before then\.

R\.

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Torsten Scheck \[mailto&#8203;:torsten\.scheck@&#8203;sot\.de\]
Sent&#8203;: Thursday\, January 04\, 2001 4&#8203;:44 AM
To&#8203;: Richard Soderberg
Subject&#8203;: Re&#8203;: \[ID 20010103\.001\] File&#8203;::Find not descending into
Joliet\-filesystem


Richard Soderberg wrote&#8203;:
 > Do you have any example code I can test on my local system
 > that would demonstrate this problem?

Sure\. Here we go&#8203;: example code & output on my RH7\-box\.

\-\-\-\-\-\-\-\-\-\-\-cut here\-\-\-\-\-\-\-\-\-\-
\#\!/usr/bin/perl \-w
use strict;
use File&#8203;::Find;
find\(sub\{print "$File&#8203;::Find&#8203;::name\\n"\}\, @&#8203;ARGV\) if @&#8203;ARGV;
\_\_END\_\_
\-\-\-\-\-\-\-\-\-\-\-cut here\-\-\-\-\-\-\-\-\-\-

\# Red Hat 7\.0

\# CD&#8203;: ISO 9660\, Rockridge\, Joliet
\#     mkisofs \-f \-L \-J \-r \-q

\# using joliet\-filesystem
mount \-t iso9660 \-o ro\,norock /dev/cdrom /mnt/mnt\\
&& \./findtest\.pl /mnt/mnt \\
&& umount /mnt/mnt

Output&#8203;:
/mnt/mnt
/mnt/mnt/README
/mnt/mnt/\_Driver
/mnt/mnt/\_Images
/mnt/mnt/\_boot
/mnt/mnt/\_pq
/mnt/mnt/\_tools
/mnt/mnt/\_win
/mnt/mnt/bin
/mnt/mnt/boot
/mnt/mnt/dev
/mnt/mnt/etc
/mnt/mnt/lib
/mnt/mnt/mnt
/mnt/mnt/proc
/mnt/mnt/sbin
/mnt/mnt/tmp
/mnt/mnt/usr
/mnt/mnt/var

\# using iso\-filesystem
mount \-t iso9660 \-o ro\,norock\,nojoliet /dev/cdrom /mnt/mnt\\
&& \./findtest\.pl /mnt/mnt \\
&& umount /mnt/mnt

Output&#8203;:
/mnt/mnt
/mnt/mnt/bin
/mnt/mnt/boot
/mnt/mnt/dev
/mnt/mnt/etc
/mnt/mnt/lib
/mnt/mnt/mnt
/mnt/mnt/proc
/mnt/mnt/readme
/mnt/mnt/sbin
/mnt/mnt/tmp
/mnt/mnt/usr
/mnt/mnt/var
/mnt/mnt/\_boot
/mnt/mnt/\_driver
/mnt/mnt/\_images
/mnt/mnt/\_pq
/mnt/mnt/\_tools
/mnt/mnt/\_win

\# using rockridge\-filesystem

mount \-t iso9660 \-o ro /dev/cdrom /mnt/mnt\\
&& \./findtest\.pl /mnt/mnt \\
&& umount /mnt/mnt

Output&#8203;:
/mnt/mnt
/mnt/mnt/mnt
/mnt/mnt/README
/mnt/mnt/tmp
/mnt/mnt/\_win
/mnt/mnt/\_win/WinME
/mnt/mnt/\_win/WinME/base2\.cab
/mnt/mnt/\_win/WinME/catalog\.cab
/mnt/mnt/\_win/WinME/cleanhd\.txt
/mnt/mnt/\_win/WinME/deltemp\.com
/mnt/mnt/\_win/WinME/dossetup\.bin
/mnt/mnt/\_win/WinME/driver5\.cab
/mnt/mnt/\_win/WinME/driver6\.cab
/mnt/mnt/\_win/WinME/driver7\.cab
/mnt/mnt/\_win/WinME/extract\.exe
/mnt/mnt/\_win/WinME/format\.com
/mnt/mnt/\_win/WinME/mini\.cab
\.
\.
\.



Richard Soderberg wrote&#8203;:

> Do you have any example code I can test on my local system that would
> demonstrate this problem?
>
> R\.
>
>
>> \-\-\-\-\-Original Message\-\-\-\-\-
>> From&#8203;: torsten@&#8203;sotlx2\.sot\.de \[mailto&#8203;:torsten@&#8203;sotlx2\.sot\.de\]
>> Sent&#8203;: Wednesday\, January 03\, 2001 2&#8203;:11 AM
>> To&#8203;: perl5\-porters@&#8203;perl\.org
>> Cc&#8203;: richard@&#8203;perl\.org
>> Subject&#8203;: \[ID 20010103\.001\] File&#8203;::Find not descending into
>> Joliet\-filesystem
>>
>>
>>
>> This is a bug report for perl from scheck@&#8203;bigfoot\.de\,
>> generated with the help of perlbug 1\.26 running under perl 5\.00503\.
>>
>> This is a bug report for perl from scheck@&#8203;bigfoot\.de\,
>> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please enter your report here\]
>> When descending into a Joliet filesystem Find stops after the
>> first level\. Rockridge or ISO CDs are ok\.
>> \[Please do not change anything below this line\]
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \-\-\-
>> Flags&#8203;:
>>     category=core
>>     severity=low
>> \-\-\-
>> Site configuration information for perl v5\.6\.0&#8203;:
>>
>> Configured by prospector at Mon Aug  7 10&#8203;:58&#8203;:30 EDT 2000\.
>>
>> Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
>>   Platform&#8203;:
>>     osname=linux\, osvers=2\.2\.5\-22smp\, archname=i386\-linux
>>     uname='linux porky\.devel\.redhat\.com 2\.2\.5\-22smp \#1 smp wed
>> jun 2 09&#8203;:11&#8203;:51 edt 1999 i686 unknown '
>>     config\_args='\-des \-Doptimize=\-O2 \-march=i386 \-mcpu=i686
>> \-Dcc=gcc \-Dcccdlflags=\-fPIC \-Dinstallprefix=/usr \-Dprefix=/usr
>> \-Darchname=i386\-linux \-Dd\_dosuid \-Dd\_semctl\_semun \-Di\_db \-Di\_ndbm
>> \-Di\_gdbm \-Di\_shadow \-Di\_syslog \-Dman3ext=3pm \-Uuselargefiles'
>>     hint=recommended\, useposix=true\, d\_sigaction=define
>>     usethreads=undef use5005threads=undef useithreads=undef
>> usemultiplicity=undef
>>     useperlio=undef d\_sfio=undef uselargefiles=undef
>>     use64bitint=undef use64bitall=undef uselongdouble=undef
usesocks=undef
>>   Compiler&#8203;:
>>     cc='gcc'\, optimize='\-O2 \-march=i386 \-mcpu=i686'\,
>> gccversion=2\.96 20000731 \(experimental\)
>>     cppflags='\-fno\-strict\-aliasing'
>>     ccflags ='\-fno\-strict\-aliasing'
>>     stdchar='char'\, d\_stdstdio=define\, usevfork=false
>>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
>>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\,
>> Off\_t='off\_t'\, lseeksize=4
>>     alignbytes=4\, usemymalloc=n\, prototype=define
>>   Linker and Libraries&#8203;:
>>     ld='gcc'\, ldflags =' \-L/usr/local/lib'
>>     libpth=/usr/local/lib /lib /usr/lib
>>     libs=\-lnsl \-ldl \-lm \-lc \-lcrypt
>>     libc=/lib/libc\-2\.1\.92\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>>   Dynamic Linking&#8203;:
>>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
>>     cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'
>>
>> Locally applied patches&#8203;:
>>
>>
>> \-\-\-
>> @&#8203;INC for perl v5\.6\.0&#8203;:
>>     /usr/lib/perl5/5\.6\.0/i386\-linux
>>     /usr/lib/perl5/5\.6\.0
>>     /usr/lib/perl5/site\_perl/5\.6\.0/i386\-linux
>>     /usr/lib/perl5/site\_perl/5\.6\.0
>>     /usr/lib/perl5/site\_perl
>>     \.
>>
>> \-\-\-
>> Environment for perl v5\.6\.0&#8203;:
>>     HOME=/home/torsten
>>     LANG=de\_DE
>>     LANGUAGE \(unset\)
>>     LD\_LIBRARY\_PATH \(unset\)
>>     LOGDIR \(unset\)
>>
>> PATH=/usr/kerberos/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin
>> :/home/torsten/bin
>>     PERL\_BADLANG \(unset\)
>>     SHELL=/bin/bash
>>
>> \-\-\-
>> Site configuration information for perl 5\.00503&#8203;:
>>
>> Configured by root at Wed Feb  2 15&#8203;:34&#8203;:50 EST 2000\.
>>
>> Summary of my perl5 \(5\.0 patchlevel 5 subversion 3\) configuration&#8203;:
>>   Platform&#8203;:
>>     osname=linux\, osvers=2\.2\.5\-22smp\, archname=i386\-linux
>>     uname='linux porky\.devel\.redhat\.com 2\.2\.5\-22smp \#1 smp wed
>> jun 2 09&#8203;:11&#8203;:51 edt 1999 i686 unknown '
>>     hint=recommended\, useposix=true\, d\_sigaction=define
>>     usethreads=undef useperlio=undef d\_sfio=undef
>>   Compiler&#8203;:
>>     cc='cc'\, optimize='\-O2 \-m486 \-fno\-strength\-reduce'\,
>> gccversion=egcs\-2\.91\.66 19990314/Linux \(egcs\-1\.1\.2 release\)
>>     cppflags='\-Dbool=char \-DHAS\_BOOL \-I/usr/local/include'
>>     ccflags ='\-Dbool=char \-DHAS\_BOOL \-I/usr/local/include'
>>     stdchar='char'\, d\_stdstdio=undef\, usevfork=false
>>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
>>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>>     alignbytes=4\, usemymalloc=n\, prototype=define
>>   Linker and Libraries&#8203;:
>>     ld='cc'\, ldflags =' \-L/usr/local/lib'
>>     libpth=/usr/local/lib /lib /usr/lib
>>     libs=\-lnsl \-ldl \-lm \-lc \-lposix \-lcrypt
>>     libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
>>   Dynamic Linking&#8203;:
>>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
>>     cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
>>
>> Locally applied patches&#8203;:
>>
>>
>> \-\-\-
>> @&#8203;INC for perl 5\.00503&#8203;:
>>     /usr/lib/perl5/5\.00503/i386\-linux
>>     /usr/lib/perl5/5\.00503
>>     /usr/lib/perl5/site\_perl/5\.005/i386\-linux
>>     /usr/lib/perl5/site\_perl/5\.005
>>     \.
>>
>> \-\-\-
>> Environment for perl 5\.00503&#8203;:
>>     HOME=/home/torsten
>>     LANG=de
>>     LANGUAGE \(unset\)
>>     LC\_ALL=de\_DE
>>     LD\_LIBRARY\_PATH=&#8203;:/opt/SYBSasa6/lib
>>     LOGDIR \(unset\)
>>
>> PATH=\.&#8203;:/usr/kerberos/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/b
>> in&#8203;:/opt/informix/bin&#8203;:/opt/SYBSasa6/bin&#8203;:/opt/kde/bin&#8203;:/usr/local/bin
>> :/mnt/d/data/projekt/mp3/bin&#8203;:/usr/local/bin&#8203;:/home/torsten/bin/vnc\_
>> x86\_linux\_2\.0
>>     PERL\_BADLANG \(unset\)
>>     SHELL=/bin/bash
>>

**Comment by p5pRT** [comment]
### From @tamias

On Tue\, Jan 09\, 2001 at 10&#8203;:11&#8203;:12PM \-0800\, Richard Soderberg wrote&#8203;:
> I don't have time to handle this until the weekend; figured I'd forward it
> along and see if someone else helps before then\.
> 
> R\.
> 
> \-\-\-\-\-Original Message\-\-\-\-\-
> From&#8203;: Torsten Scheck \[mailto&#8203;:torsten\.scheck@&#8203;sot\.de\]
> Sent&#8203;: Thursday\, January 04\, 2001 4&#8203;:44 AM
> To&#8203;: Richard Soderberg
> Subject&#8203;: Re&#8203;: \[ID 20010103\.001\] File&#8203;::Find not descending into
> Joliet\-filesystem
> 
> 
> Richard Soderberg wrote&#8203;:
>  > Do you have any example code I can test on my local system
>  > that would demonstrate this problem?
> 
> Sure\. Here we go&#8203;: example code & output on my RH7\-box\.
> 
> \-\-\-\-\-\-\-\-\-\-\-cut here\-\-\-\-\-\-\-\-\-\-
> \#\!/usr/bin/perl \-w
> use strict;
> use File&#8203;::Find;
> find\(sub\{print "$File&#8203;::Find&#8203;::name\\n"\}\, @&#8203;ARGV\) if @&#8203;ARGV;
> \_\_END\_\_
> \-\-\-\-\-\-\-\-\-\-\-cut here\-\-\-\-\-\-\-\-\-\-


Try setting $File&#8203;::Find&#8203;::dont\_use\_nlink\.


Ronald

**Comment by p5pRT** [comment]
### From @floatingatoll

R\.

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Torsten Scheck \[mailto&#8203;:torsten\.scheck@&#8203;sot\.de\]
Sent&#8203;: Wednesday\, January 10\, 2001 8&#8203;:43 AM
To&#8203;: Richard Soderberg; Ronald J Kimball
Subject&#8203;: Re&#8203;: \[ID 20010103\.001\] File&#8203;::Find not descending into
Joliet\-filesystem


 > >  > \-\-\-\-\-\-\-\-\-\-\-cut here\-\-\-\-\-\-\-\-\-\-
 > > \#\!/usr/bin/perl \-w
 > > use strict;
 > > use File&#8203;::Find;
 > > find\(sub\{print "$File&#8203;::Find&#8203;::name\\n"\}\, @&#8203;ARGV\) if @&#8203;ARGV;
 > > \_\_END\_\_
 > > \-\-\-\-\-\-\-\-\-\-\-cut here\-\-\-\-\-\-\-\-\-\-

 >  Try setting $File&#8203;::Find&#8203;::dont\_use\_nlink\.
 > Ronald

It works\! Thank you\.

Why do I have to set it anyway? Shouldn't this be the default 
behaviour of find? Would I get into trouble\, if I was always using it?

I've searched perldoc and the web but there wasn't a more detailed 
description of $File&#8203;::Find&#8203;::dont\_use\_nlink\.

Torsten

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> From&#8203;: Torsten Scheck \[mailto&#8203;:torsten\.scheck@&#8203;sot\.de\]

> It works\! Thank you\.

> Why do I have to set it anyway? Shouldn't this be the default behaviour
> of find? Would I get into trouble\, if I was always using it?

> I've searched perldoc and the web but there wasn't a more detailed
> description of $File&#8203;::Find&#8203;::dont\_use\_nlink\.

It's slightly slower\, but it will always work\.  It's necessary to set this
variable when using any file system that doesn't use Unix semantics for
nlink on directories \(AFS has the same problem\)\.

**Comment by p5pRT** [comment]
### From @floatingatoll

If this is document in the docs of File&#8203;::Find\, the bug report can be closed\.
Thanks\, Russ\.

R\.

> \-\-\-\-\-Original Message\-\-\-\-\-
> From&#8203;: Russ Allbery \[mailto&#8203;:rra@&#8203;stanford\.edu\]
> Sent&#8203;: Wednesday\, January 10\, 2001 3&#8203;:50 PM
> To&#8203;: Richard Soderberg
> Cc&#8203;: perl5\-porters@&#8203;perl\.org; torsten\.scheck@&#8203;sot\.de
> Subject&#8203;: Re&#8203;: \[ID 20010103\.001\] File&#8203;::Find not descending into
> Joliet\-filesystem
>
>
> > From&#8203;: Torsten Scheck \[mailto&#8203;:torsten\.scheck@&#8203;sot\.de\]
>
> > It works\! Thank you\.
>
> > Why do I have to set it anyway? Shouldn't this be the default behaviour
> > of find? Would I get into trouble\, if I was always using it?
>
> > I've searched perldoc and the web but there wasn't a more detailed
> > description of $File&#8203;::Find&#8203;::dont\_use\_nlink\.
>
> It's slightly slower\, but it will always work\.  It's necessary to set this
> variable when using any file system that doesn't use Unix semantics for
> nlink on directories \(AFS has the same problem\)\.
>
> \-\-
> Russ Allbery \(rra@&#8203;stanford\.edu\)             \<http&#8203;://www\.eyrie\.org/~eagle/>
>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-

Moin\,

run across this some weeks ago\, and was biten by it today\, while reading
Simons summaries\. So I dived into it&#8203;:

The problem is that stat\(\) returns 1 for dirs on such a CD\-ROM\, instead of
subdircount\+2\. The following patch works around this\. I have no idea wether
stat\, the filesystem or Perl is broken\, or wether this is a feature of
these CD\-ROMs\.

Cheers\,

Tels


\- \-\-\- Find\.pm\.orig     Tue Nov 14 20&#8203;:27&#8203;:57 2000
\+\+\+ Find\.pm     Fri Jan 12 17&#8203;:22&#8203;:17 2001
@&#8203;@&#8203; \-465\,6 \+465\,10 @&#8203;@&#8203;
&nbsp;        @&#8203;filenames = readdir DIR;
&nbsp;        closedir\(DIR\);
 
\+       \# on Joliet ISO CD\-ROMs stat\(\) returns falsely 1 as nlink count
\+       \# instead subdirs\+2\, so we work around&#8203;:
\+       $avoid\_nlink = 1 if $nlink \< 2;
\+
&nbsp;        if \($nlink == 2 && \!$avoid\_nlink\) \{
&nbsp;            \# This dir has no subdirectories\.
&nbsp;            for my $FN \(@&#8203;filenames\) \{


Regards\,

Tels

\- \-\-
 "Why do you go so slowly? Do you think this is some kind of game?"
 http&#8203;://bloodgate\.com/thief/     Thief \- The Dark Project
 http&#8203;://bloodgate\.com/aifilter   Rewriting the HTML as we know it\.
 http&#8203;://freedomforlinks\.de       Fight for your right to link\. 
 PGP key available on http&#8203;://bloodgate\.com/tels\.asc or via email\.

The email address contained in the From/Reply of this email is NOT
to be displayed on webpages\, entered into any mailing lists\, given 
to other people\, used to send me electronic greeting cards without
my written\, signed permission\. No SPAM nor UCE\, either\! Thank you\!
And PLEASE refrain from forwarding me chain letters\, massmailings\,
or pyramid schemes\.

\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: 2\.6\.3i
Charset&#8203;: latin1

iQEVAwUBOl8wxncLPEOTuEwVAQHCvAf/TYS5ACXVhkA9rR3DpVv/vDdyIvVD5lsp
1ztK\+ezAffO\+9B0tkqTvsAVOtpxAFfPG9aopsvIS6lG6pjBZ0zAHT5zTq3\+j/ApQ
QWOC/k0cfNnSzDFVjOuGrFQiMzZZ5FTO2nsTaIlBDL8ij3UDOIhJGNYHEIsk5NUD
ktYcmegM9g8jBA3qWJi1cA30IEUsnrxwTgGD1WwFHdr2SzYaArwszFOFWVLvEMOk
zWEOaN64ZRAfuCpFF1o8TRa2h3nHP8p5czbeBt52DRQyq9m2TwRCb9KBE8984YTI
oovNea9HfK3DTOHXK5x8MWSRobMBMmKWr440InCgRvdHV/lQdurmtw==
=jCgQ
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-

Moin\,

On 12\-Jan\-01 Tels tried to scribble about&#8203;:
> \-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-
> \- \-\-\- Find\.pm\.orig     Tue Nov 14 20&#8203;:27&#8203;:57 2000
> \+\+\+ Find\.pm     Fri Jan 12 17&#8203;:22&#8203;:17 2001
\[snip\]

Actually\, here is an even better patch\. The previous one switched to
slow\-mode upon hitting the first dir without proper stat\. This one
preserves the global setting and goes faster when coming out of the
equivalent of /cdrom again\. Probably some doc fixes \(mentioning joliet/afs\)
would be needed as well? Can anyone test this withFS please? I'll try an
OS/2/NFS test on monday\.

> The email address contained in the From/Reply of this email is NOT

ARG\, I thought I set up a rule to strip this for p5p\.\.\.sorry\.

Cheers\,

Tels

\- \-\-\- /usr/local/lib/perl5/5\.6\.0/File/Find\.pm     Tue Nov 14 20&#8203;:27&#8203;:57 2000
\+\+\+ Find\.pm     Fri Jan 12 17&#8203;:51&#8203;:15 2001
@&#8203;@&#8203; \-393\,7 \+393\,7 @&#8203;@&#8203;
&nbsp;     my \($CdLvl\,$Level\) = \(0\,0\);
&nbsp;     my @&#8203;Stack;
&nbsp;     my @&#8203;filenames;
\- \-    my \($subcount\,$sub\_nlink\);
\+    my \($subcount\,$sub\_nlink\,$do\_avoid\_nlink\);
&nbsp;     my $SE= \[\];
&nbsp;     my $dir\_name= $p\_dir;
&nbsp;     my $dir\_pref= \( $p\_dir eq '/' ? '/' : "$p\_dir/" \);
@&#8203;@&#8203; \-465\,7 \+465\,13 @&#8203;@&#8203;
&nbsp;        @&#8203;filenames = readdir DIR;
&nbsp;        closedir\(DIR\);
 
\- \-       if \($nlink == 2 && \!$avoid\_nlink\) \{
\+       \# on Joliet ISO CD\-ROMs stat\(\) returns falsely 1 as nlink count
\+       \# instead subdirs\+2\, so we work around\. Use slow method only for the
\+       \# problematic cases&#8203;:
\+       $do\_avoid\_nlink = $avoid\_nlink;
\+       $do\_avoid\_nlink = 1 if $nlink \< 2;
\+
\+       if \($nlink == 2 && \!$do\_avoid\_nlink\) \{
&nbsp;            \# This dir has no subdirectories\.
&nbsp;            for my $FN \(@&#8203;filenames\) \{
&nbsp;                next if $FN =~ /^\\\.\{1\,2\}\\z/;
@&#8203;@&#8203; \-482\,7 \+488\,7 @&#8203;@&#8203;
 
&nbsp;            for my $FN \(@&#8203;filenames\) \{
&nbsp;                next if $FN =~ /^\\\.\{1\,2\}\\z/;
\- \-               if \($subcount > 0 || $avoid\_nlink\) \{
\+               if \($subcount > 0 || $do\_avoid\_nlink\) \{
&nbsp;                    \# Seen all the subdirs?
&nbsp;                    \# check for directoriness\.
&nbsp;                    \# stat is faster for a file in the current directory



Tels

\- \-\-
 "Why do you go so slowly? Do you think this is some kind of game?"
 http&#8203;://bloodgate\.com/thief/     Thief \- The Dark Project
 http&#8203;://bloodgate\.com/aifilter   Rewriting the HTML as we know it\.
 http&#8203;://freedomforlinks\.de       Fight for your right to link\. 
 PGP key available on http&#8203;://bloodgate\.com/tels\.asc or via email\.


\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: 2\.6\.3i
Charset&#8203;: latin1

iQEVAwUBOl82xHcLPEOTuEwVAQGLagf/U1lfRJQlsyppQlXwcovWcRPnVPG6JRHV
ojFDsi1EYBY18r2a3atyrfNYi0q6jJ1DJc67RbMzkrxFc1IucEcC3ZNj1N3VaftO
2RKB6HOlzOPqfawE9o0EVk5AdkAoEjac6yrQ0Q4CDccVenFH1nVJ\+\+mI\+5ZOlEw2
5IGh29CM/gCMbzCmMNcWTVRHJ4I8yk/4oz5qh\+aqWjgONori1VqIqM1ZTVf5nPIp
B2uuVU/MGZHE2WZfgmHS6s9sqAkUmQIipkM\+kxqKivGBWic8DnwCdjZ984YJHH92
arkqaq9GzkcCQV8b/rUpj9wDD\+wm\+7P7ZAIRrpwHbW6SUb3mSoW1Ng==
=Zfxh
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-

**Comment by p5pRT** [comment]
### From @floatingatoll



\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Torsten Scheck \[mailto&#8203;:torsten\.scheck@&#8203;sot\.de\]
Sent&#8203;: Monday\, January 15\, 2001 4&#8203;:00 AM
To&#8203;: Richard Soderberg
Subject&#8203;: Re&#8203;: FW&#8203;: \[PATCH\]\[ID 20010103\.001\] File&#8203;::Find not descending into
Joli


Richard Soderberg wrote&#8203;:

> Can you test this patch?  It should solve your problem\.

I'll try\.

Besides&#8203;: I tried to use File&#8203;::Find to generate playlists for Hardware 
MP3\-Player\. As these devices often use the ISO\-filesystem\, Linux with 
his "nojoliet\, norock" mount option is very good for this kind of 
jobs\. But the way File&#8203;::Find traverses the filesystem isn't 
predictable\. Perl 5\.6's File&#8203;::Find returns all folders even in reverse 
order\, whereas older versions under perl 5\.004 are doing this right\.
So at the moment I'm doing these things with a simple and small \(but 
unoptimized\) recursive function\. Is it planned to support different 
ways of traversing the directory tree? \(except for the finddepth\, 
which just insures listing the contents of a directory first\)

Torsten

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


> So at the moment I'm doing these things with a simple and small \(but 
> unoptimized\) recursive function\. Is it planned to support different 
> ways of traversing the directory tree? \(except for the finddepth\, 
> which just insures listing the contents of a directory first\)
> 
> Torsten
> 
> 

File&#8203;::Find with bleadperl supports C\<preprocess> which allows sorting 
of directory entries\.

Robin

**Comment by p5pRT** [comment]
### From @gannett-ggreer

The included patch was applied in 5fa2bf2b4e1a8 on October 19\, 2001\.

\-\- 
George Greer

**Comment by p5pRT** [comment]
@gannett-ggreer - Status changed from 'open' to 'resolved'
